apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/components/template-selector-control.js

```diff
@@ -1,64 +1,116 @@
 /**
  * External dependencies
  */
-import { isEmpty } from 'lodash';
+import { isEmpty, each } from 'lodash';
 import classnames from 'classnames';
 
 /**
  * WordPress dependencies
  */
-import { withInstanceId } from '@wordpress/compose';
+import { withInstanceId, compose } from '@wordpress/compose';
 import { BaseControl } from '@wordpress/components';
+import { memo, Component } from '@wordpress/element';
+import { parse as parseBlocks } from '@wordpress/blocks';
 
-function TemplateSelectorControl( {
-	label,
-	className,
-	help,
-	instanceId,
-	onClick,
-	templates = [],
-} ) {
-	const id = `template-selector-control-${ instanceId }`;
-	const handleButtonClick = event => onClick( event.target.value );
-
-	if ( isEmpty( templates ) ) {
-		return null;
+/**
+ * Internal dependencies
+ */
+import TemplateSelectorItem from './template-selector-item';
+import replacePlaceholders from '../utils/replace-placeholders';
+
+// Load config passed from backend.
+const { siteInformation = {} } = window.starterPageTemplatesConfig;
+
+class TemplateSelectorControl extends Component {
+	state = {
+		blocks: {},
+	};
+
+	constructor( props ) {
+		super();
+
+		// Populate the state with the parsed blocks,
+		// only in the dynamic preview mode.
+		if ( props.templates && props.dynamicPreview ) {
+			const blocks = {};
+			each( props.templates, ( { content, slug } ) => {
+				if ( content ) {
+					blocks[ slug ] = parseBlocks( content );
+				}
+			} );
+			this.state.blocks = blocks;
```

Comment:
I don't think we should be writing state based on props here. It breaks the data flow. If we pass an updated prop the state never gets recomputed as the constructor only runs once.

See https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html

Reply:
good point, you're right.
I've changed the approach again, using `useMemo()` and storing the parsed content into the `<TemplateSelectorPreview >` component.